design-docs/005-manifest-split.md

```diff
@@ -0,0 +1,406 @@
+                                          # in the future we'll implement initialized-chunks
+                                          # that will be useful for sparse arrays
+
+      target: coord1                      # arrays that match will go to this manifest set
+      
+
+    - metadata-chunks: [0, 200]           # arrays < 200 chunks will go to coord2
+                                          # but only if they didn't get assigned by the rule above
+      target: coord2
+
+
+    - metadata-chunks: [2000000, null]    # huge arrays have their special manifest set
+      target: big-array
+
+
+  preload:                      # what manifests to asynchronously preload on session start
+    max-manifest-size: 50000    # don't preload manifests larger than this
+    max-manifests: 10           # don't preload more than this many manifests
+    arrays:                     # currently we support selecting by path only
+                                # we will refine with more power if needed
+
+      - path: .*/time           # will preload manifests for all arrays matching this regex
+      - path: .*/latitude
+      - path: .*/longitude
+```
+
+The full power of the configuration is there for users that really need it. But
+usually the default configuration will be enough.
+
+Here is the default configuration:
+
+```yaml
+chunk-manifests:  # of course, we'll have to tune all these numbers
+  sets:
+    - coordinates:                # we intend coordinate arrays to be assigned here
+        max-manifest-size: 50000
+        cardinality: 1
+        max-arrays-per-manifest: null 
+        overflow-to: default
+
+    - default:
+        max-manifest-size: 1000000
+        cardinality: null
+        max-arrays-per-manifest: null
+        overflow-to: null
+
+  rules:
+    - metadata-chunks: [0, 5000]
+      path: .*
+      target: coordinates
+
+
+  # should we actually do this by default or is no-preload a better default?
+  preload:
+    max-manifest-size: 50000
+    max-manifests: 1
+    arrays:
+      - path: .*/time
+      - path: .*/latitude
+      - path: .*/longitude
+```
+
+### Algorithm
+
+Notice there are two different methods for packing, one based on
+`max-manifest-size` and the other on `arrays-per-manifest`. We may chose to
+implement the second one at a later stage.
+
+Configuration is always validated before saving it:
+  * type checks
+  * `target` exists
+  * no `overflow-to` loops
+  * one of `max-manifest-size` or `arrays-per-manifest`
+  * default cannot have cardinality
+  * etc.
+
+#### Packing by `max-manifest-size`
+
+We define the closure of a set of arrays. In plain English, the closure of a
+set of arrays `A` is the set of all arrays that would need their manifest
+rewritten if we update all arrays in `A`. We are going to re-pack all arrays
+in the closure of the set of modified arrays.
+
+Example:
+
+Manifest 1 has arrays: a, b
+Manifest 2 has arrays: a, b, c
+Manifest 3 has arrays: c
```

Comment:
Feels like we'd gain a large amount of simplicity by disallowing that kind of packing.

For example, 
```
Manifest 1 has arrays: a, b
Manifest 2 has arrays: a, b
Manifest 3 has arrays: a  # b is over; c is not allowed
Manifest 4 has arrays: c
Manifest 5 has arrays: c
```

I'm not sure we gain a lot by allowing the  packing of `a,c` in Manifest 3 (assuming that matches the rules). Indeed we might be better off relaxing the threshold to fit `a` between 1,2 by being looser about enforcing `max-manifest-size`. Something like if the overflow of `a` in Manifest 3 is <20% of `max-manifest-size` then just redistribute over preceding manifests.